feat: AutoRod

sqlerrorthing opened the pull request:
Auto use fishing rod to PVP.
---
The original code was taken from [LiquidBounce Legacy](https://github.com/CCBlueX/LiquidBounce/blob/legacy/src/main/java/net/ccbluex/liquidbounce/features/module/modules/combat/AutoRod.kt).
 
Coded by @zyklone4096 (in https://github.com/CCBlueX/LiquidBounce/pull/930)
Ported & recoded by @sqlerrorthing

closes #3693
closes #4397
closes #5526
closes #5114

sqlerrorthing committed:
feat/auto-rod

sqlerrorthing committed:
implementation

sqlerrorthing committed:
refactor

sqlerrorthing committed:
fixes

sqlerrorthing committed:
fixes

sqlerrorthing committed:
idk

sqlerrorthing committed:
working!

sqlerrorthing committed:
mini refactoring

sqlerrorthing committed:
module description

sqlerrorthing committed:
padding

sqlerrorthing committed:
useless code

sqlerrorthing committed:
useless code

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/FacingEnemy.kt:
```diff
@@ -0,0 +1,38 @@
+
+    private inline val nearbyEnemies: List<Entity>
+        get() = world.entities.filter {
+            it.shouldBeAttacked() && it.pos.squaredDistanceTo(player.pos) <= activationDistance.random()
```
`activationDistance.random().sq()`

sqlerrorthing replied:
> `activationDistance.random().sq()`

ill resolve it when i can

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,68 @@
+import net.ccbluex.liquidbounce.utils.inventory.Slots
+import net.minecraft.item.Items
+
+@get:JvmSynthetic
```
why?

MukjepScarlet replied:
This annotation means to hide it in Java side

sqlerrorthing replied:
bc the method / getter is inlined.

> This annotation means to hide it in Java side

i know.

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,68 @@
+@file:Suppress("MaxLineLength")
```
Reviewers won't like it :p

sqlerrorthing replied:
bc long link in kdoc.
https://github.com/CCBlueX/LiquidBounce/blob/ce9a862f1b7dc911c7432e1851992dd30e4b6b97/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt#L22

sqlerrorthing committed:
squared activation distance

sqlerrorthing committed:
sync rotations

sqlerrorthing commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,68 @@
+@file:Suppress("MaxLineLength")
```
bc long link in kdoc.
https://github.com/CCBlueX/LiquidBounce/blob/ce9a862f1b7dc911c7432e1851992dd30e4b6b97/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt#L22

sqlerrorthing committed:
useless condition

sqlerrorthing committed:
destructuring

sqlerrorthing committed:
refactoring

sqlerrorthing committed:
useless annotations

sqlerrorthing committed:
more suffixes

sqlerrorthing committed:
ms to ticks

MukjepScarlet commented on src/main/resources/resources/liquidbounce/lang/ru_ru.json:
```diff
@@ -627,6 +627,7 @@
   "liquidbounce.module.holeFiller.description": "Автоматически заполняет безопасные места для PvP с кристаллами.",
   "liquidbounce.module.tpAura.description": "Автоматически телепортирует и атакует врагов вокруг.",
   "liquidbounce.module.aspect.description": "Позволяет растягивать или сужать игру по горизонтали.",
+  "liquidbounce.module.autoRod.description": "Автоматически использует удочку при пвп.",
```
In other items it's `PvP` instead of `пвп`

sqlerrorthing replied:
Pvp - Игрок (Player) Против (Versus) Игрока (Player)
`ПвП` - This is a slang expression, tacitly accepted in our Russian Minecraft community. Плеер (Player) Версус (Versus) Плеер (Player)
but nobody used is as `ПвП`, using `Пвп` or `пвп`.

`пвп` is literally just a transcription of `pvp`

sqlerrorthing committed:
russian translation fix

sqlerrorthing committed:
moved the conditions

sqlerrorthing committed:
.

1zun4 commented:
So, I have tried the Auto Rod on Gamster Classic - but it barely works, it pulls back before it reaches the target. It also does not switch back to the sword without the use of Auto Weapon. I don't know.

superblaubeere27 commented:
There is another PR which implements this better. I am working on this other PR.

sqlerrorthing committed:
fix stop using rod

sqlerrorthing committed:
refactoring & pullback when reached

sqlerrorthing commented:
> So, I have tried the Auto Rod on Gamster Classic - but it barely works, it pulls back before it reaches the target. It also does not switch back to the sword without the use of Auto Weapon. I don't know.

> it pulls back before it reaches the target

added the setting `OnTargetReach`

> It also does not switch back to the sword

maybe i fixed that in https://github.com/CCBlueX/LiquidBounce/pull/5798/commits/97f838303007445b00caa8fa435eac6fe727841b (no), in https://github.com/CCBlueX/LiquidBounce/pull/5798/commits/3f7a0b485ec2f1e5870db2623a572712bcf45713?

sqlerrorthing committed:
swap back

sqlerrorthing commented:
it's swapping slots back before release the rod, idk why
https://github.com/CCBlueX/LiquidBounce/blob/34ada88c885134b60b1bf0b36259dbc01f1bc2ba/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Using.kt#L49-L56

sqlerrorthing commented:
> There is another PR which implements this better. I am working on this other PR.

where?

sqlerrorthing committed:
reset chronomenters

sqlerrorthing committed:
refactor

sqlerrorthing committed:
useless condition

sqlerrorthing committed:
refactor

sqlerrorthing committed:
refactor

sqlerrorthing committed:
health by scoreboard

sqlerrorthing committed:
Merge branch 'nextgen' into feat/auto-rod

# Conflicts:
#	src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/elytratarget/AutoFirework.kt
#	src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt

SuicidalEeper commented:
Is this dead :c

sqlerrorthing commented:
> Is this dead :c

maybe

SuicidalEeper commented:
> > Is this dead :c
> 
> maybe

Crying rn

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/FacingEnemy.kt:
```diff
@@ -0,0 +1,43 @@
+            it.shouldBeAttacked() && it.pos.squaredDistanceTo(player.pos) <= activationDistance.random().sq()
+        }
+
+    @Suppress("ReturnCount", "NOTHING_TO_INLINE")
```
`NOTHING_TO_INLINE`.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/FacingEnemy.kt:
```diff
@@ -0,0 +1,43 @@
+    private val ignoreOnLowHealth by boolean("IgnoreOnLowHealth", true)
+    private val healthThreshold by int("HealthThreshold", 10, 1..20, suffix = "hp")
+
+    private inline val nearbyEnemies: List<Entity>
```
Noting to inline.

And this should be a function, not a property.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/FacingEnemy.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+
+    @Suppress("ReturnCount", "NOTHING_TO_INLINE")
+    internal inline fun testUseRod(healthByScoreboard: Boolean): Boolean {
```
Nothing to inline.

Missing Documentation

sqlerrorthing replied:
> Missing Documentation

Im too lazy to wrote it

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,59 @@
+
+    private val healthByScoreboard by boolean("HealthByScoreboard", true)
+
+    private inline val usingRod
```
Nothing to inline.

This could be an extension of the player

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,59 @@
+                && player.activeItem?.item == Items.FISHING_ROD)
+                || using.isUsingRod
+
+    private inline val testFacingEnemyUseRod
```
Nothing to inline.

A function, not a property.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/ModuleAutoRod.kt:
```diff
@@ -0,0 +1,59 @@
+                && player.getActualHealth(healthByScoreboard) >= playerHealthThreshold.toFloat()
+                && facingEnemy.testUseRod(healthByScoreboard)
+
+    private inline val canUseRod get() = when {
```
Nothing to inline.

And this is not a property, it is a function.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Pullback.kt:
```diff
@@ -0,0 +1,40 @@
+    private val onTargetReach by boolean("OnTargetReach", true)
+    private val delay by int("Pullback", 15, 1..20, suffix = "ticks")
+
+    private inline val fishingBobberEntity
```
Nothing to inline.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Pullback.kt:
```diff
@@ -0,0 +1,40 @@
+            else -> world.entities.find { it is FishingBobberEntity && it.owner == player } as FishingBobberEntity?
+        }
+
+    private inline val targetReached
```
Noting to inline - not a property

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Pullback.kt:
```diff
@@ -0,0 +1,40 @@
+            else -> fishingBobberEntity?.takeIf { it.hookedEntity != null } != null
+        }
+
+    internal inline fun testPullbackRod(pullback: () -> Unit) {
```
A good function name and documentation is missing.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Push.kt:
```diff
@@ -0,0 +1,24 @@
+internal class Push : Configurable("Push") {
+    private val delay by int("Delay", 2, 1..20, suffix = "ticks")
+
+    internal inline fun testPushRod(push: () -> Unit) {
+        if (pushChronometer.hasElapsed(delay.toLong() * MILLISECONDS_PER_TICK)) {
+            push()
+            reset()
+        }
+    }
```
s.a.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Using.kt:
```diff
@@ -0,0 +1,62 @@
+    internal var isUsingRod = false
+    private var resetSlot: Int? = null
+
+    internal inline fun startRodUsing(slot: HotbarItemSlot) = push.testPushRod {
```
Nothing to inline

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/autorod/Using.kt:
```diff
@@ -0,0 +1,62 @@
+        push.reset()
+    }
+
+    @get:JvmSynthetic
+    internal inline val canUseRodWhenUsingItem
+        get() = onItemUsing
+                || (player.activeItem?.item != Items.FISHING_ROD
+                    && !(player.usingItem || KillAuraAutoBlock.blockVisual))
```
I think you know my opinion on that by now

MukjepScarlet replied:
In my opinion `inline` has 3 usages:
1. functional parameters like `List.map`
2. `<reified T>`
3. very simple body like `public actual inline fun min(a: Double, b: Double): Double = nativeMath.min(a, b)`

superblaubeere27 replied:
@MukjepScarlet 

1. Agreed
2. Agreed
3. No! Never ever do this! Even if the function has a simple body. It takes information away from the JIT compiler which can hurt performance. For example if a function mostly returns *false* or a specific value, the JIT compiler can reason about it in all places this function is used. If it is inlined, the JIT compiler has no idea that those code snippets are correlated. Even if it does not hurt performance I think that adding stuff which is useless at best for no reason should be avoided.

superblaubeere27 reviewed (changes requested):
Please stop using inline if you are not inlining a functional or generic parameter. It is useless at best, increases file sizes and potentially hurts performance.

sqlerrorthing committed:
code refactor

1zun4 committed:
Merge branch 'nextgen' into fork/sqlerrorthing/feat/auto-rod

# Conflicts:
#	src/main/resources/resources/liquidbounce/lang/ru_ru.json

1zun4 commented:
The module still barely or does not work at all.

1zun4 committed:
fix: merge

1zun4 commented:
I have closed the pull request, as the module is written in a complicated way and performs badly. There is little that can be done to improve it besides rewriting it from scratch.

The issues I mentioned months ago are still there and it also fails to hit the target the entire time.
